### PR TITLE
Project: Fake Quantization 

### PR DESCRIPTION
**Pull request by ranbir-sharma**
***Project : Fake Quantization***

Project Details are as follow

- Fake Quantization is added to the LLTFI to ensure fault injection within the targeted layers specified in the yaml files 
- Lib code is added to facilitate the process for maintaining and Quantizing / Dequantizing the values as provided 
- CMake Files are updated to build the code for Fake Quantization while setting up LLTFI

### CODE REVIEW

**Comment by abrahamchan** [comment]
Do you have any tests for this feature? For example, it would be great to have an example under [sample_programs/ml_sample_programs/](https://github.com/DependableSystemsLab/LLTFI/tree/master/sample_programs/ml_sample_programs/) to demonstrate the use of fake quantization.

**Comment by ranbir-sharma** [comment]
> Do you have any tests for this feature? For instance, it would be great to have an example under [sample_programs/ml_sample_programs/](https://github.com/DependableSystemsLab/LLTFI/tree/master/sample_programs/ml_sample_programs/) to demonstrate the use of fake quantization.
> 
> Also, the [input_masterlist_ml.yaml](https://github.com/DependableSystemsLab/LLTFI/blob/master/docs/input_masterlist_ml.yaml) should be updated with the `fakeQuant` option.

Added Documentation in the `input_masterlist_ml.yaml` with examples and added error checks for fake Quantization in `instrument.py`

File: llvm_passes/core/FakeQuantizationPass.cpp [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+#include <cxxabi.h>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "llvm/Support/CommandLine.h"
+// #include "Controller.h"
```
Comment (abrahamchan) [inline llvm_passes/core/FakeQuantizationPass.cpp:20]: Please delete this include if not required.

**Review by abrahamchan — CHANGES_REQUESTED** [review]

File: llvm_passes/core/FakeQuantizationPass.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+void findTotalLayers(llvm::Module &M);
+void insertFakeQuantForInjectFault(llvm::Function &F);
+void insertFakeQuantForProfiling(llvm::Function &F);
+void insertFakeQuantInst(llvm::Module &M, bool IsProfiling);
+std::string getOperandName(llvm::Instruction *I);
+
+} // namespace llfi
+
+#endif // LLFI_H
```
[the remark this answers is not in the document]
Comment (ranbir-sharma) [reply llvm_passes/core/FakeQuantizationPass.h]: Resolved!

File: bin/instrument.py [hunk 3]
```diff
@@ -317,6 +319,26 @@ def readCompileOption():
...
+    
+    assert isinstance(targetLayer, str), "TargetLayer must be of type string" 
+    assert targetLayer == "conv" or targetLayer == "matmul" , "TargetLayer can only be 'conv' and 'matmul'" 
+    assert isinstance(minPercentileOutlierThreshold, int), "Minimum Percentile Value should be of type int"
+    assert isinstance(maxPercentileOutlierThreshold, int), "Maximum Percentile Value should be of type int"
+    assert isinstance(bitWidth, int), "BitWidth should be of type int"
+    assert 0 < bitWidth, "BitWidth must be a integer greater than 0"
+    assert math.log2(bitWidth).is_integer(), "BitWidth must be a exponent of power 2"
+    assert 0 <= minPercentileOutlierThreshold and minPercentileOutlierThreshold <= 100, "Minimum Percentile Value for Percentile should be greater than or equal to 0"
+    assert 0 <= maxPercentileOutlierThreshold and maxPercentileOutlierThreshold <= 100, "Maximum Percentile Value for Percentile should be greater than or equal to 0 and lesser than or equal to 100"
```
[the remark this answers is not in the document]
Comment (ranbir-sharma) [reply bin/instrument.py]: Resolved!

File: runtime_lib/FakeQuantizationLib.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+float Quantize(float w1, float x1, int currentLayerIndex, int totalNumberOfLayers);
+float QuantizeMatMul(float w1, float x1, int currentLayerIndex, int totalNumberOfLayers);
+float FakeQuantIntegerBasedAddition(float num1, float num2);
+float FakeQunatDequnatizeAndBiasAddition(float num1, float num2);
+void getBias(float a, float b);
+#ifdef __cplusplus
+}
+#endif
+
+#endif // FAKE_QUANTIZATION_H
```
[the remark this answers is not in the document]
Comment (ranbir-sharma) [reply runtime_lib/FakeQuantizationLib.h]: Resolved!

File: llvm_passes/core/FakeQuantizationPass.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (ranbir-sharma) [reply llvm_passes/core/FakeQuantizationPass.h]: Resolved!

File: runtime_lib/FakeQuantizationLib.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+#include <stdio.h>
+#include <math.h>
+#include <limits.h>
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+float getWAndX(float w1, float x1, int currentLayerIndex, int totalNumberOfLayers);
+void finished(int currentLayerIndex, int totalNumberOfLayers, int minPercentileThreshold, int maxPercetileThreshold, int bitWidth);
```
[the remark this answers is not in the document]
Comment (ranbir-sharma) [reply runtime_lib/FakeQuantizationLib.h]: Resolved!

File: runtime_lib/FakeQuantizationLib.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+float dequantize(int q);
+float Quantize(float w1, float x1, int currentLayerIndex, int totalNumberOfLayers);
+float QuantizeMatMul(float w1, float x1, int currentLayerIndex, int totalNumberOfLayers);
+float FakeQuantIntegerBasedAddition(float num1, float num2);
+float FakeQunatDequnatizeAndBiasAddition(float num1, float num2);
```
[the remark this answers is not in the document]
Comment (ranbir-sharma) [reply runtime_lib/FakeQuantizationLib.h]: Resolved!

File: docs/input_masterlist_ml.yaml [hunk 7]
```diff
@@ -139,3 +139,25 @@ runOption:
...
+      targetLayer: conv
+      minPercentileOutlierThreshold : 10
+      maxPercentileOutlierThreshold : 90
+      bitWidth: 8
+      
+
+# Fake Qunatization parameters comes within compileOption. In here, you need to mention the targer layer - conv or matmul (currently only supporting these)
+# minPercentileOutlierThreshold and maxPercentileOutlierThreshold are optional parameters. On not mentioning these parameters, they are autoamtically set to 
+# minPercentileOutlierThreshold = 0 and maxPercentileOutlierThreshold = 100
+# bitWidth needs to be mentioned and is the bit width of the Quantized Int numbers which needs to be a exponent of power 2
```
[the remark this answers is not in the document]
Comment (ranbir-sharma) [reply docs/input_masterlist_ml.yaml]: Resolved!

File: docs/input_masterlist_ml.yaml [hunk 8]
```diff
@@ -139,3 +139,25 @@ runOption:
...
+
+compileOption:
+    fakeQuant: 
+      targetLayer: conv
+      minPercentileOutlierThreshold : 10
+      maxPercentileOutlierThreshold : 90
+      bitWidth: 8
+      
+
+# Fake Qunatization parameters comes within compileOption. In here, you need to mention the targer layer - conv or matmul (currently only supporting these)
```
[the remark this answers is not in the document]
Comment (ranbir-sharma) [reply docs/input_masterlist_ml.yaml]: Resolved!

File: llvm_passes/core/FakeQuantizationPass.cpp (hunk 1 above)
Comment (ranbir-sharma) [reply llvm_passes/core/FakeQuantizationPass.cpp]: Resolved!

File: runtime_lib/FakeQuantizationLib.cpp [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+        {
+            ans = max_number;
+        }
+        if (ans < (-1 * max_number))
+        {
+            ans = (-1 * max_number);
+        }
+        // printf("This is going away - %i with w - %f with Q(w) - %i and x - %f with Q(x) - %i\n", ans, w1, quantize_helper(w1, s_w), x1, quantize_helper(x1, s_x));
+        return ans;
+        // return 8.0;
```
Comment (abrahamchan) [inline runtime_lib/FakeQuantizationLib.cpp:324]: Remove this code.

File: runtime_lib/FakeQuantizationLib.cpp [hunk 10]
```diff
@@ -0,0 +1,427 @@
...
```
Comment (abrahamchan) [inline runtime_lib/FakeQuantizationLib.cpp:305]: Remove this code.

File: runtime_lib/FakeQuantizationLib.cpp [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+    if (maxPercentileThreshold == 100)
+    {
+        max_w = w_values[w_index - 1];
+    }
+    else
+    {
+        max_w = find_percentile_value(w_values, w_index, maxPercentileThreshold);
+    }
+
+    // min_w = find_percentile_value(w_values, w_index, lowerPercentile);
```
Comment (abrahamchan) [inline runtime_lib/FakeQuantizationLib.cpp:211]: Remove this section of commented out code.

File: runtime_lib/FakeQuantizationLib.cpp [hunk 12]
```diff
@@ -0,0 +1,427 @@
...
+
+    // return (int)(w1 * x1);
+    return (w1 * x1);
+}
+
+float FakeQuantIntegerBasedAddition(float num1, float num2)
+{
+    int intNum1 = (int)num1;
+    int intNum2 = (int)num2;
+    // printf("Got this in Addition - %i and %i and their sum %i\n", intNum1, intNum2, intNum1 + intNum2);
```
Comment (abrahamchan) [inline runtime_lib/FakeQuantizationLib.cpp:401]: Remove this debugging log.

File: runtime_lib/FakeQuantizationLib.cpp [hunk 13]
```diff
@@ -0,0 +1,427 @@
...
+        int ans = (quantize_helper(w1, s_w) * quantize_helper(x1, s_x));
+        if (ans > (max_number))
+        {
+            ans = max_number;
+        }
+        if (ans < (-1 * max_number))
+        {
+            ans = (-1 * max_number);
+        }
+        // printf("This is going away - %i with w - %f with Q(w) - %i and x - %f with Q(x) - %i\n", ans, w1, quantize_helper(w1, s_w), x1, quantize_helper(x1, s_x));
```
Comment (abrahamchan) [inline runtime_lib/FakeQuantizationLib.cpp:322]: Remove this debugging log.

File: runtime_lib/FakeQuantizationLib.cpp [hunk 14]
```diff
@@ -0,0 +1,427 @@
...
+    // printf("Got this in Addition - %i and %i and their sum %i\n", intNum1, intNum2, intNum1 + intNum2);
+    return (float)(intNum1 + intNum2);
+}
+
+float FakeQuantDequantizeAndBiasAddition(float QunatizeNum, float Basis)
+{
+    int QunatizeIntNum = (int)QunatizeNum;
+    float dequnatizedFloat = dequantize(QunatizeIntNum);
+    float result = dequnatizedFloat + Basis;
+    // printf("Returning this - %f\n", result);
```
Comment (abrahamchan) [inline runtime_lib/FakeQuantizationLib.cpp:410]: Remove these prints used for debugging.

**Review by abrahamchan — CHANGES_REQUESTED** [review]
Could you please make a second round of changes? I didn't catch the other typos in the previous review.

**Comment by ManiSadati** [comment]
Does this PR still have the unexpected large value issue for all the models?

**Comment by abrahamchan** [comment]
LGTM!

**Pull request merged**